test/NFTDescriptor.spec.ts

```diff
@@ -609,56 +647,202 @@ describe('NFTDescriptor', () => {
 \\nPool Address: ${poolAddress}\\n${quoteTokenSymbol} Address: ${quoteTokenAddress.toLowerCase()}\\n${baseTokenSymbol} Address: ${baseTokenAddress.toLowerCase()}\\n\
-Fee Tier: ${fee}\\nToken ID: ${tokenId}\\n\\n⚠️ DISCLAIMER: Due diligence is imperative when assessing this NFT. Make sure token addresses match the expected tokens, as \
-token symbols may be imitated.", "image": "${encodedSvgImage(quoteTokenAddress, baseTokenAddress)}"}`
+Fee Tier: ${feeTier}\\nToken ID: ${tokenId}\\n\\n⚠️ DISCLAIMER: Due diligence is imperative when assessing this NFT. Make sure token addresses match the expected tokens, as \
+token symbols may be imitated.", "image": "${await encodedSvgImage(
+      tokenId,
+      feeTier,
+      quoteTokenAddress,
+      baseTokenAddress,
+      quoteTokenSymbol,
+      baseTokenSymbol,
+      overRange,
+      tickLower,
+      tickUpper
+    )}"}`
+  }
+
+  async function svgImage(
+    tokenId: number,
+    feeTier: string,
+    quoteTokenAddress: string,
+    baseTokenAddress: string,
+    quoteTokenSymbol: string,
+    baseTokenSymbol: string,
+    overRange: number,
+    tickLower: number,
+    tickUpper: number
+  ): Promise<string> {
+    const tokenToNum = (tokenAddr: string, index: number): number => {
+      return parseInt(tokenAddr.slice(index, index + 2), 16)
+    }
+    const scale = (n: number, inMn: number, inMx: number, outMn: number, outMx: number): number => {
+      return Math.floor(((n - inMn) * (outMx - outMn)) / (inMx - inMn) + outMn)
+    }
+    baseTokenAddress = baseTokenAddress.toLowerCase()
+    quoteTokenAddress = quoteTokenAddress.toLowerCase()
+    const fade = overRange === -1 ? '#fade-up' : overRange === 1 ? '#fade-down' : '#none'
+    const curve = await nftDescriptor.getCurve(tickLower, tickUpper)
+    const curveCircle = await nftDescriptor.generateSVGCurveCircle(overRange)
+    const color0 = tokenToColorHex(quoteTokenAddress, 2)
+    const color1 = tokenToColorHex(baseTokenAddress, 2)
+    const color2 = tokenToColorHex(quoteTokenAddress, 36)
+    const color3 = tokenToColorHex(baseTokenAddress, 36)
+    const x1 = scale(tokenToNum(quoteTokenAddress, 36), 0, 255, 16, 274)
+    const y1 = scale(tokenToNum(baseTokenAddress, 36), 0, 255, 100, 484)
+    const x2 = scale(tokenToNum(quoteTokenAddress, 32), 0, 255, 16, 274)
+    const y2 = scale(tokenToNum(baseTokenAddress, 32), 0, 255, 100, 484)
+    const x3 = scale(tokenToNum(quoteTokenAddress, 28), 0, 255, 16, 274)
+    const y3 = scale(tokenToNum(baseTokenAddress, 28), 0, 255, 100, 484)
+    const str1length = 7 * (tokenId.toString().length + 4 + 4)
+    const str2length = 7 * (tickLower.toString().length + 5 + 4)
+    const str3length = 7 * (tickUpper.toString().length + 5 + 4)
+    return `<svg width="290" height="500" viewBox="0 0 290 500" xmlns="http://www.w3.org/2000/svg" xmlns:xlink='http://www.w3.org/1999/xlink'><style>@import \
```

Comment:
soo....I did this. Was it the best way to do this?? I have no idea.

Reply:
yes you're right. this is actually a shit show 😝. I do need to test some of the text but I can parse the little snippets of the JSON that i Need to test.

Reply:
if you want to validate the svg, i would parse the svg and check its elements